<h1>Tabla de Contenidos<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Conceptos-importantes" data-toc-modified-id="Conceptos-importantes-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Conceptos importantes</a></span></li><li><span><a href="#El-dataset-que-vamos-a-utilizar" data-toc-modified-id="El-dataset-que-vamos-a-utilizar-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>El <em>dataset</em> que vamos a utilizar</a></span></li><li><span><a href="#¿Qué-es-el-EDA?" data-toc-modified-id="¿Qué-es-el-EDA?-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>¿Qué es el EDA?</a></span><ul class="toc-item"><li><span><a href="#¿Cómo-realizar-un-análisis-exploratorio-de-datos?" data-toc-modified-id="¿Cómo-realizar-un-análisis-exploratorio-de-datos?-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>¿Cómo realizar un análisis exploratorio de datos?</a></span></li></ul></li><li><span><a href="#Proceso-de-EDA-sobre-los-datos" data-toc-modified-id="Proceso-de-EDA-sobre-los-datos-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Proceso de EDA sobre los datos</a></span><ul class="toc-item"><li><span><a href="#Entender-las-variables/columnas-del-dataset" data-toc-modified-id="Entender-las-variables/columnas-del-dataset-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Entender las variables/columnas del <em>dataset</em></a></span></li><li><span><a href="#Valores-nulos" data-toc-modified-id="Valores-nulos-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>Valores nulos</a></span><ul class="toc-item"><li><span><a href="#Métodos-para-identificar-nulos-en-Pandas" data-toc-modified-id="Métodos-para-identificar-nulos-en-Pandas-4.2.1"><span class="toc-item-num">4.2.1&nbsp;&nbsp;</span>Métodos para identificar nulos en Pandas</a></span></li></ul></li><li><span><a href="#Valores-duplicados" data-toc-modified-id="Valores-duplicados-4.3"><span class="toc-item-num">4.3&nbsp;&nbsp;</span>Valores duplicados</a></span><ul class="toc-item"><li><span><a href="#Métodos-para-indentificar-los-duplicados-en-Pandas" data-toc-modified-id="Métodos-para-indentificar-los-duplicados-en-Pandas-4.3.1"><span class="toc-item-num">4.3.1&nbsp;&nbsp;</span>Métodos para indentificar los duplicados en Pandas</a></span></li></ul></li></ul></li><li><span><a href="#Resumen-de-los-conceptos-aprendidos" data-toc-modified-id="Resumen-de-los-conceptos-aprendidos-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Resumen de los conceptos aprendidos</a></span></li><li><span><a href="#Ejercicios" data-toc-modified-id="Ejercicios-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Ejercicios</a></span></li></ul></div>

# Conceptos importantes 

En este apartado vamos a definir algunos de los conceptos que más repetiremos a lo largo de todo el módulo y que es importante que tengamos claros:

- **Variables**:  una variable se refiere a una característica o atributo específico que se está midiendo, registrando o estudiando en un conjunto de datos. Puede ser cualquier cantidad, calidad o propiedad que pueda variar entre diferentes elementos o unidades de observación en un estudio. Las variables pueden clasificarse en diferentes tipos, y el tipo de variable determina las operaciones que se pueden realizar con ella y las formas en que se puede analizar. En nuestro caso, cada variable será cada columna de nuestro *DataFrame*. 

- **Tipos de variables**: Es importante comprender los tipos de variables presentes en un conjunto de datos, ya que pueden afectar el tipo de análisis y las técnicas que se pueden aplicar a los datos.

    - **Numéricas**: Estas variables representan valores numéricos y se utilizan para medir magnitudes. En el conjunto de datos, la variable "age" (edad) es un ejemplo de una variable numérica.

    - **Categóricas**: Estas variables representan categorías o etiquetas discretas. En el conjunto de datos, las variables "job" (tipo de trabajo), "marital" (estado civil), "education" (educación)-, "default" (crédito en incumplimiento), "housing" (préstamo hipotecario), "loan" (préstamo personal), "contact" (tipo de comunicación de contacto), "month" (mes), "day_of_week" (día de la semana) y "poutcome" (resultado de la campaña anterior) son ejemplos de variables categóricas.Dentro de las variables categóricas, podemos tener varios tipos dependiendo de la naturaleza de los datos. A continuación, se presentan algunos tipos comunes de variables categóricas:

        - Nominales: Son variables categóricas sin un orden inherente entre las categorías. Las categorías representan diferentes grupos o clases, pero no hay una relación de jerarquía entre ellas. Ejemplos de variables nominales en el conjunto de datos proporcionado podrían ser "job" (tipo de trabajo) y "marital" (estado civil).

        - Ordinales: Son variables categóricas con un orden o jerarquía entre las categorías. Las categorías representan diferentes niveles o rangos, y el orden entre ellas es significativo. Ejemplos de variables ordinales en el conjunto de datos podrían ser "education" (educación), donde las categorías pueden indicar niveles educativos ascendentes (por ejemplo, 'basic.4y', 'basic.6y', 'basic.9y', 'high.school', 'professional.course', 'university.degree'), o "poutcome" (resultado de la campaña anterior), donde las categorías pueden ser 'failure', 'nonexistent' y 'success'.

        - Binarias: Son un caso especial de variables categóricas que toman solo dos categorías posibles. Las categorías se pueden representar como '0' y '1', 'no' y 'yes', o cualquier otra combinación de dos valores opuestos. Un ejemplo de variable binaria en el conjunto de datos es "default" (crédito en incumplimiento), que puede tener los valores 'no', 'yes' o 'unknown'.


-  **Estadística Descriptiva**: Es el proceso de resumir y organizar datos para comprender sus características principales. Incluye medidas de tendencia central (media, mediana, moda) y medidas de dispersión (varianza, desviación estándar).

- **Estadística Inferencial**: Se utiliza para hacer inferencias o predicciones sobre una población a partir de una muestra de datos. Incluye pruebas de hipótesis, intervalos de confianza y regresión.

-  **Análisis Exploratorio de Datos (EDA)**: El proceso de examinar y visualizar datos para identificar patrones, tendencias y posibles valores atípicos. Gráficos, diagramas de dispersión y resúmenes estadísticos son herramientas comunes en EDA.

- **Validación y Limpieza de Datos**: La validación implica asegurarse de que los datos sean precisos y confiables, mientras que la limpieza implica eliminar valores atípicos, datos duplicados o inconsistencias en los datos..

- **Ética en el Análisis de Datos**: Comprender las implicaciones éticas de la recopilación y el uso de datos es crucial. Esto incluye la privacidad de los datos y la responsabilidad en el manejo de la información. Os compartimos [aquí](https://www.linkedin.com/pulse/la-%C3%A9tica-en-el-an%C3%A1lisis-de-datos-desaf%C3%ADos-y-consideraciones/?originalSubdomain=es) y [aquí](https://www.puromarketing.com/101/34076/etica-datos-marketing-digital) un par de artículos interesantes sobre este tema. 


# El *dataset* que vamos a utilizar

Los datos están relacionados con campañas de *marketing* directo de una institución bancaria portuguesa. Las campañas de marketing se basaron en llamadas telefónicas. A menudo, se requería más de un contacto con el mismo cliente para determinar si el producto (depósito a plazo bancario) sería suscrito o no. Las colummas que tenemos en este *dataset* son: 

En base al dataset proporcionado, a continuación se presenta una breve descripción de cada una de las columnas:

- **age**: La edad del cliente.

- **job**: La ocupación o profesión del cliente.

- **marital**: El estado civil del cliente.

- **education**: El nivel educativo del cliente.

- **default**: Indica si el cliente tiene algún historial de incumplimiento de pagos (1: Sí, 0: No).

- **housing**: Indica si el cliente tiene un préstamo hipotecario (1: Sí, 0: No).

- **loan**: Indica si el cliente tiene algún otro tipo de préstamo (1: Sí, 0: No).

- **contact**: El método de contacto utilizado para comunicarse con el cliente.

- **duration**: La duración en segundos de la última interacción con el cliente.

- **campaign**: El número de contactos realizados durante esta campaña para este cliente.

- **emp.var.rate**: La tasa de variación del empleo.

- **cons.price.idx**: El índice de precios al consumidor.

- **cons.conf.idx**: El índice de confianza del consumidor.

- **euribor3m**: La tasa de interés de referencia a tres meses.

- **nr.employed**: El número de empleados.

- **y**: Indica si el cliente ha suscrito un producto o servicio (Sí/No).

- **date**: La fecha en la que se realizó la interacción con el cliente.

- **latitude**: La latitud de la ubicación asociada al cliente.

- **longitude**: La longitud de la ubicación asociada al cliente.

- **id_**: Un identificador único para cada registro en el dataset.


# ¿Qué es el EDA?

En esta primera lección nos centraremos en el proceso de EDA, pero nos podríamos preguntar que significa esto "del EDA". Imaginemos que estamos con un grupo de amigos y nos dicen de ir a comer a un restaurante del que han oído hablar. Sin embargo, no te dicen nada más, no nos dicen qué tipo de comida sirven, cómo es el restaurante, el precio medio de la comida etc. En esta situación nos podemos ver algo desconcertadas con un montón de preguntas a las que queremos dar una respuesta para tomar una decisión.

En esta situación, la primera pregunta que nos podríamos hacer es: ¿qué tipo de comida sirven? o incluso ver algunas opiniones previas de otros clientes. Es decir, haremos un trabajo de investigación para saber si nos interesa ir o no ir a ese restaurante. Esto no es más que lo que hacen los analistas de datos, solo que nosotras lo llamaremos **Análisis Exploratorio de Datos** o **EDA**. 


El **Análisis Exploratorio de Datos** (EDA por sus siglas en inglés *Exploratory Data Analysis*) es una etapa fundamental en el proceso de análisis de datos. Consiste en examinar y visualizar los datos de manera sistemática con el objetivo de descubrir patrones, identificar tendencias, detectar valores atípicos y extraer información relevante para comprender mejor el conjunto de datos.

El EDA se realiza generalmente al principio de un proyecto de análisis de datos, antes de aplicar técnicas de limpieza más avanzadas. A través del análisis exploratorio, se busca familiarizarse con los datos, identificar posibles problemas de calidad, determinar la distribución y variabilidad de las variables, y generar hipótesis que puedan guiar el análisis posterior.

Algunas de las técnicas comunes utilizadas en el Análisis Exploratorio de Datos incluyen:

1. Resumen estadístico: Se calculan medidas descriptivas como la media, la mediana, la desviación estándar, el rango y los percentiles para comprender la distribución y variabilidad de los datos.

2. Visualización de datos: Se utilizan gráficos y diagramas, como histogramas, gráficos de dispersión, diagramas de caja, diagramas de barras y gráficos de línea, para visualizar las características de los datos y explorar las relaciones entre las variables.

3. Análisis de correlación: Se examina la correlación entre las variables para identificar posibles relaciones y dependencias. Esto se puede realizar mediante el cálculo de coeficientes de correlación, como el coeficiente de correlación de Pearson.

4. Tratamiento de valores atípicos: Se buscan valores atípicos o extremos que puedan tener un impacto significativo en los resultados del análisis. Estos valores pueden ser errores de medición o indicativos de fenómenos inusuales o interesantes.

5. Imputación de datos faltantes: Si hay datos faltantes en el conjunto de datos, se pueden utilizar técnicas para estimar o imputar los valores faltantes de manera que no afecten significativamente los resultados del análisis.

6. Segmentación y agrupación: Se pueden aplicar técnicas de segmentación o agrupación para identificar subgrupos o patrones particulares dentro del conjunto de datos.

> En resumen, el Análisis Exploratorio de Datos se refiere al proceso de realizar una serie de investigaciones inciales sobre los datos que tenemos para poder descubrir patrones, detectar anomalías, probar hipótesis y comprobar suposiciones con la ayuda de estadísticas y representaciones gráficas. 

Y es que es **ESENCIAL ENTENDER LOS DATOS** y tratar de **OBTENER el mayor número de CONOCIMIENTO** a partir de ellos. 

Sin embargo, es importante tener en cuenta que el análisis exploratorio de datos no es un conjunto de técnicas, pasos o reglas. **El análisis exploratorio es una "filosofía" sobre como se deben analizar los datos** dependiendo en gran parte de las preguntas que queramos responder. 

## ¿Cómo realizar un análisis exploratorio de datos?

Como hemos dicho, el EDA no tiene una receta universal que podamos seguir paso a paso. Entonces nos puede surgir la siguiente pregunta:

> **¿Cómo realizar un análisis exploratorio de datos?**
>
>En realidad, esta es una de las preguntas que todo el mundo desea conocer. 

La respuesta es que depende del conjunto de datos con el que se trabaje. No hay un método o unos métodos comunes para llevar a cabo EDA. 

A lo largo de esta lección vamos a tratar de entender algunos métodos comunes y los pasos que se utilizan en el proceso de EDA. Estas "pasos" podrían ser: 

- **Entender las variables/columnas**: en esta parte entenderemos qué variables tenemos, qué tipos de datos, si tenemos valores nulos o duplicados, qué valores únicos tenemos en nuestro *dataset*, etc.

- **Limpiar el *dataset***: una vez identificados todos los valores únicos, los duplicados y/o nulos procederemos a dejar nuestro *dataset* lo más limpio posible, es decir, quitar duplicados, nulos, columnas irrelevantes etc. 

- **Analizar las relaciones entre las variables**: nos permitirá encontrar patrones, relaciones o anomalías en nuestros datos que nos ayudarán a sacar conclusiones. 

# Proceso de EDA sobre los datos

## Entender las variables/columnas del *dataset*

Se refiere a la tarea de investigar y comprender las columnas presentes en el conjunto de datos que estamos analizando. Algunas de las tareas que podríamos incluir en esta fase son:

1. **Identificar las variables**: En esta etapa, se examina el conjunto de datos para identificar todas las variables o columnas presentes. Se puede revisar la descripción del conjunto de datos o explorar los encabezados de las columnas para comprender qué aspectos se están registrando o midiendo.

2. **Tipos de datos**: Es importante comprender qué tipos de datos contiene cada variable. Algunas variables pueden ser numéricas, otras categóricas, binarias, ordinales o de tiempo, como se mencionó anteriormente. Esto es relevante para seleccionar las técnicas adecuadas de análisis y visualización posteriormente.

3. **Valores nulos o faltantes**: Durante esta etapa, se busca la presencia de valores nulos o faltantes en las variables. Los valores faltantes pueden surgir debido a errores en la recopilación de datos, problemas técnicos o simplemente porque algunos elementos no tienen valores registrados para ciertas variables. La identificación de valores faltantes es crucial para decidir cómo manejarlos posteriormente, ya sea eliminando filas con valores faltantes o realizando una imputación de datos para llenar los vacíos.

4. **Valores duplicados**: Se busca la presencia de valores duplicados en las variables. Esto implica verificar si hay registros idénticos o muy similares en las columnas del conjunto de datos. Los valores duplicados pueden afectar los análisis y generar resultados incorrectos, por lo que es importante detectarlos y decidir cómo manejarlos, ya sea eliminándolos o investigando su causa.

5. **Valores únicos**: Se analizan los valores únicos presentes en cada variable. Esto puede revelar información importante sobre la variabilidad y la distribución de los datos. Contar los valores únicos y su frecuencia puede ayudar a identificar patrones o anomalías, y proporcionar una visión general de la diversidad de los datos.

En resumen, al "entender las variables/columnas" durante el Análisis Exploratorio de Datos, nos enfocamos en examinar y comprender las características y propiedades de las variables presentes en el conjunto de datos. Esto nos permite obtener información inicial sobre la estructura y la calidad de los datos antes de realizar análisis más detallados y tomar decisiones sobre cómo manejar los datos faltantes, duplicados o únicos.

In [2]:
# Tratamiento de datos
# -----------------------------------------------------------------------
import pandas as pd

# Configuración
# -----------------------------------------------------------------------
pd.set_option('display.max_columns', None) # para poder visualizar todas las columnas de los DataFrames

In [3]:
# cargamos el csv con el que vamos a ir desarrollando la lección de hoy, recordad que la descripción la tenéis al inicio del jupyter. 
# para eso usaremos el método "pd.read_csv" que ya vimos al inicio del módulo 2 cuando aprendimos las herramientas de web scrapping
df = pd.read_csv("../files/bank-additional.csv", index_col = 0)

# hacemos un ".head()" para mostrar las 5 primeras filas del dataframe
df.head()

age        job  marital    education  default  housing  loan    contact  \
0   NaN  housemaid  MARRIED     basic.4y      0.0      0.0   0.0  telephone   
1  57.0   services  MARRIED  high.school      NaN      0.0   0.0  telephone   
2  37.0   services  MARRIED  high.school      0.0      1.0   0.0  telephone   
3  40.0     admin.  MARRIED     basic.6y      0.0      0.0   0.0  telephone   
4  56.0   services  MARRIED  high.school      0.0      0.0   1.0  telephone   

   duration  campaign  ...  emp.var.rate  cons.price.idx cons.conf.idx  \
0       261         1  ...           1.1          93,994         -36,4   
1       149         1  ...           1.1          93,994         -36,4   
2       226         1  ...           1.1          93,994         -36,4   
3       151         1  ...           1.1          93,994         -36,4   
4       307         1  ...           1.1          93,994         -36,4   

   euribor3m nr.employed   y                date latitude longitude  \
0      4,857        5191  no       2-agosto-2019   41.495   -71.233   
1        NaN        5191  no  14-septiembre-2016   34.601   -83.923   
2      4,857        5191  no     15-febrero-2019   34.939   -94.847   
3        NaN        5191  no   29-noviembre-2015   49.041   -70.308   
4        NaN        5191  no       29-enero-2017   38.033  -104.463   

                                    id_  
0  089b39d8-e4d0-461b-87d4-814d71e0e079  
1  e9d37224-cb6f-4942-98d7-46672963d097  
2  3f9f49b5-e410-4948-bf6e-f9244f04918b  
3  9991fafb-4447-451a-8be2-b0df6098d13e  
4  eca60b76-70b6-4077-80ba-bc52e8ebb0eb  

[5 rows x 23 columns]

In [4]:
# oh oh... si nos fijamos bien, no podemos ver todas las columnas del DataFrame (hay '...' en mitad de nuestra tabla) 
# ¿qué podemos hacer para conseguir que se vean todas las gráficas? 
pd.set_option('display.max_columns', None)

# después de aplicar el método ya podremos ver todas las columnas del df.
df.head()


age        job  marital    education  default  housing  loan    contact  \
0   NaN  housemaid  MARRIED     basic.4y      0.0      0.0   0.0  telephone   
1  57.0   services  MARRIED  high.school      NaN      0.0   0.0  telephone   
2  37.0   services  MARRIED  high.school      0.0      1.0   0.0  telephone   
3  40.0     admin.  MARRIED     basic.6y      0.0      0.0   0.0  telephone   
4  56.0   services  MARRIED  high.school      0.0      0.0   1.0  telephone   

   duration  campaign  pdays  previous     poutcome  emp.var.rate  \
0       261         1    999         0  NONEXISTENT           1.1   
1       149         1    999         0  NONEXISTENT           1.1   
2       226         1    999         0  NONEXISTENT           1.1   
3       151         1    999         0  NONEXISTENT           1.1   
4       307         1    999         0  NONEXISTENT           1.1   

  cons.price.idx cons.conf.idx euribor3m nr.employed   y                date  \
0         93,994         -36,4     4,857        5191  no       2-agosto-2019   
1         93,994         -36,4       NaN        5191  no  14-septiembre-2016   
2         93,994         -36,4     4,857        5191  no     15-febrero-2019   
3         93,994         -36,4       NaN        5191  no   29-noviembre-2015   
4         93,994         -36,4       NaN        5191  no       29-enero-2017   

   latitude  longitude                                   id_  
0    41.495    -71.233  089b39d8-e4d0-461b-87d4-814d71e0e079  
1    34.601    -83.923  e9d37224-cb6f-4942-98d7-46672963d097  
2    34.939    -94.847  3f9f49b5-e410-4948-bf6e-f9244f04918b  
3    49.041    -70.308  9991fafb-4447-451a-8be2-b0df6098d13e  
4    38.033   -104.463  eca60b76-70b6-4077-80ba-bc52e8ebb0eb

In [5]:
# recordemos que también tenemos el método ".tail()", el cual nos mostrará las últimas 5 filas del dataframe
df.tail()

age          job  marital            education  default  housing  \
19154   NaN       admin.  MARRIED    university.degree      0.0      0.0   
26206  34.0   technician  MARRIED  professional.course      0.0      1.0   
15046   NaN  blue-collar   SINGLE             basic.6y      0.0      1.0   
15280   NaN       admin.  MARRIED    university.degree      NaN      0.0   
27570   NaN   unemployed   SINGLE    university.degree      0.0      0.0   

       loan   contact  duration  campaign  pdays  previous     poutcome  \
19154   0.0  cellular       618         2    999         0  NONEXISTENT   
26206   1.0  cellular        42         7    999         0  NONEXISTENT   
15046   0.0  cellular       391         2    999         0  NONEXISTENT   
15280   0.0  cellular       674         3    999         0  NONEXISTENT   
27570   1.0  cellular       104         2    999         0  NONEXISTENT   

       emp.var.rate cons.price.idx cons.conf.idx euribor3m nr.employed    y  \
19154           1.4         93,444         -36,1       NaN      5228,1  yes   
26206          -0.1           93,2           -42       NaN      5195,8   no   
15046           1.4         93,918         -42,7       NaN      5228,1   no   
15280           1.4         93,918         -42,7     4,958      5228,1   no   
27570          -0.1           93,2           -42     4,021      5195,8   no   

                     date  latitude  longitude  \
19154     13-octubre-2015    38.147   -105.582   
26206       17-marzo-2018    49.235   -112.201   
15046  15-septiembre-2016    40.679   -120.015   
15280  23-septiembre-2019    27.772   -117.518   
27570    6-noviembre-2019    41.146   -105.026   

                                        id_  
19154  4eed05de-2a98-4227-b488-32122009b638  
26206  0f0aca88-4088-4fe2-905f-44fb675d9493  
15046  cadadd4b-7ee5-4019-b13a-ca01bb67ca5b  
15280  5f432048-d515-4bb5-9c94-62db451f88d4  
27570  993bbbd6-4dbc-4a40-a408-f91f8462bee6

In [6]:
# a veces puede que no nos interese mostrar las 5 primeras o 5 últimas filas de un dataframe, y puede que solo queramos mostrar las dos primeras o las dos últimas
# esto lo podemos hacer también con ".head()" o ".tail()", incluyendo el número de filas que queremos ver dentro del paréntesis

df.head(2) # en este caso mostramos solo las dos primeras filas

age        job  marital    education  default  housing  loan    contact  \
0   NaN  housemaid  MARRIED     basic.4y      0.0      0.0   0.0  telephone   
1  57.0   services  MARRIED  high.school      NaN      0.0   0.0  telephone   

   duration  campaign  pdays  previous     poutcome  emp.var.rate  \
0       261         1    999         0  NONEXISTENT           1.1   
1       149         1    999         0  NONEXISTENT           1.1   

  cons.price.idx cons.conf.idx euribor3m nr.employed   y                date  \
0         93,994         -36,4     4,857        5191  no       2-agosto-2019   
1         93,994         -36,4       NaN        5191  no  14-septiembre-2016   

   latitude  longitude                                   id_  
0    41.495    -71.233  089b39d8-e4d0-461b-87d4-814d71e0e079  
1    34.601    -83.923  e9d37224-cb6f-4942-98d7-46672963d097

In [7]:
# o solo la últimas 3 filas
df.tail(3)

age          job  marital          education  default  housing  loan  \
15046  NaN  blue-collar   SINGLE           basic.6y      0.0      1.0   0.0   
15280  NaN       admin.  MARRIED  university.degree      NaN      0.0   0.0   
27570  NaN   unemployed   SINGLE  university.degree      0.0      0.0   1.0   

        contact  duration  campaign  pdays  previous     poutcome  \
15046  cellular       391         2    999         0  NONEXISTENT   
15280  cellular       674         3    999         0  NONEXISTENT   
27570  cellular       104         2    999         0  NONEXISTENT   

       emp.var.rate cons.price.idx cons.conf.idx euribor3m nr.employed   y  \
15046           1.4         93,918         -42,7       NaN      5228,1  no   
15280           1.4         93,918         -42,7     4,958      5228,1  no   
27570          -0.1           93,2           -42     4,021      5195,8  no   

                     date  latitude  longitude  \
15046  15-septiembre-2016    40.679   -120.015   
15280  23-septiembre-2019    27.772   -117.518   
27570    6-noviembre-2019    41.146   -105.026   

                                        id_  
15046  cadadd4b-7ee5-4019-b13a-ca01bb67ca5b  
15280  5f432048-d515-4bb5-9c94-62db451f88d4  
27570  993bbbd6-4dbc-4a40-a408-f91f8462bee6

In [8]:
# ¿y si quisieramos ver "x" filas al azar de nuestro conjunto de datos? Podemos usar el método ".sample()", el cual por defecto nos devuelve solo una fila
df.sample()

age       job marital education  default  housing  loan   contact  \
30878  37.0  services  SINGLE  basic.9y      0.0      0.0   0.0  cellular   

       duration  campaign  pdays  previous     poutcome  emp.var.rate  \
30878        37         1    999         0  NONEXISTENT          -1.8   

      cons.price.idx cons.conf.idx euribor3m nr.employed   y          date  \
30878         92,893         -46,2       NaN      5099,1  no  5-junio-2017   

       latitude  longitude                                   id_  
30878    25.861   -117.221  68ac3007-2b7b-4d32-802c-aabc523e635e

In [9]:
# si quisieramos ver 10 filas al azar de nuestro conjunto de datos solo tendríamos que especificarlo dentro de los paréntesis
df.sample(10)

age          job   marital            education  default  housing  \
542    57.0   technician   MARRIED  professional.course      NaN      0.0   
39006   NaN      retired   MARRIED             basic.4y      0.0      NaN   
27178  57.0   management  DIVORCED  professional.course      0.0      1.0   
4161   47.0   technician   MARRIED  professional.course      NaN      1.0   
34603   NaN   technician   MARRIED  professional.course      0.0      0.0   
13605  38.0  blue-collar    SINGLE             basic.6y      0.0      1.0   
12395  31.0     services   MARRIED          high.school      NaN      0.0   
3085   45.0  blue-collar   MARRIED             basic.9y      NaN      0.0   
2491   54.0       admin.   MARRIED    university.degree      0.0      0.0   
28084  32.0  blue-collar   MARRIED             basic.9y      0.0      0.0   

       loan    contact  duration  campaign  pdays  previous     poutcome  \
542     1.0  telephone       289         1    999         0  NONEXISTENT   
39006   NaN  telephone       137         8    999         0  NONEXISTENT   
27178   0.0   cellular       849         1    999         0  NONEXISTENT   
4161    0.0  telephone       197         4    999         0  NONEXISTENT   
34603   0.0   cellular       239         1    999         1      FAILURE   
13605   1.0   cellular        95         1    999         0  NONEXISTENT   
12395   0.0   cellular       160         3    999         0  NONEXISTENT   
3085    0.0  telephone        90         3    999         0  NONEXISTENT   
2491    0.0  telephone       172         2    999         0  NONEXISTENT   
28084   0.0   cellular        66         2    999         0  NONEXISTENT   

       emp.var.rate cons.price.idx cons.conf.idx euribor3m nr.employed    y  \
542             1.1         93,994         -36,4     4,857        5191   no   
39006          -3.0         92,713           -33     0,715      5023,5   no   
27178          -0.1           93,2           -42     4,021      5195,8  yes   
4161            1.1         93,994         -36,4       NaN        5191   no   
34603          -1.8         92,893         -46,2     1,266      5099,1   no   
13605           1.4         93,918         -42,7       NaN      5228,1   no   
12395           1.4         93,918         -42,7       NaN      5228,1   no   
3085            1.1         93,994         -36,4       NaN        5191   no   
2491            1.1         93,994         -36,4     4,856        5191   no   
28084          -1.8         93,075         -47,1       NaN      5099,1   no   

                     date  latitude  longitude  \
542       31-octubre-2018    30.430   -122.688   
39006       27-julio-2019    37.870   -100.433   
27178   13-noviembre-2019    35.724    -73.894   
4161       13-agosto-2018    29.212    -74.818   
34603        6-abril-2015    35.502    -83.294   
13605  27-septiembre-2018    43.739   -118.950   
12395  24-septiembre-2016    43.238   -110.603   
3085    25-noviembre-2018    37.470   -103.193   
2491        6-agosto-2015    26.311   -107.040   
28084   24-noviembre-2018    30.910   -122.466   

                                        id_  
542    7130c4f5-8a03-4fe3-b5b0-98308ecf3427  
39006  1cfae0ff-d442-44d8-8c32-abe759d3ae13  
27178  fb675f5c-29a3-4f5a-ab41-3f250583fb42  
4161   79e0697b-645f-4295-8282-37120bf4a67d  
34603  54c522f6-7c38-4d0c-bf48-b0cf9dc9a2ef  
13605  9cef39dd-9462-49ff-911a-0e2d1879cd9d  
12395  cdd7434c-eeaf-496b-959b-3221773a5b17  
3085   cfd85c8b-2602-4661-b4ef-80bf3d7ecc37  
2491   f7d06e37-cdd3-4536-b989-3678569edfc9  
28084  8c1ab92f-f3be-4b40-995b-8b1da80c0331

In [10]:
# recordemos otro método que aprendimos al inicio del módulo 2, el método".shape" 
# que nos permitía saber cuántas filas y columnas tenemos en el DataFrame. 
# 📌 Recordemos que nos devuelve una tupla!

print(f"El número de filas que tenemos es {df.shape[0]}, y el número de columnas es {df.shape[1]}")


El número de filas que tenemos es 43000, y el número de columnas es 23


In [11]:
# otro de los métodos que más usaremos es el método que nos devuelve el nombre de todas las columnas de nuestro DataFrame. Este es el método '.columns' 
# nos va a devolver un array con todos los nombres de las columnas
df.columns

Index(['age', 'job', 'marital', 'education', 'default', 'housing', 'loan',
       'contact', 'duration', 'campaign', 'pdays', 'previous', 'poutcome',
       'emp.var.rate', 'cons.price.idx', 'cons.conf.idx', 'euribor3m',
       'nr.employed', 'y', 'date', 'latitude', 'longitude', 'id_'],
      dtype='object')

Sigamos aprendiendo nuevos métodos de Pandas que nos pueden ayudar en este análisis exploratorio de los datos, el método `.describe()`. Es una herramienta esencial para obtener una visión general de las estadísticas descriptivas de un DataFrame, lo que facilita la comprensión de la distribución de datos. Este método proporciona una instantánea concisa de las principales características numéricas de las columnas, como la media, la desviación estándar, los valores mínimo y máximo, los percentiles y más.

In [32]:
#utilizamos el me´todo '.describe()' para sacar los principales estadísticos del DataFrame
# por defecto, nos devolverá solo los estadísticos de las VARIABLES NUMÉRICAS

df.describe()

age       default       housing          loan      duration  \
count  37880.000000  34019.000000  41974.000000  41974.000000  43000.000000   
mean      39.977112      0.000088      0.535998      0.155620    257.739279   
std       10.437957      0.009390      0.498708      0.362499    258.666033   
min       17.000000      0.000000      0.000000      0.000000      0.000000   
25%       32.000000      0.000000      0.000000      0.000000    102.000000   
50%       38.000000      0.000000      1.000000      0.000000    179.000000   
75%       47.000000      0.000000      1.000000      0.000000    319.000000   
max       98.000000      1.000000      1.000000      1.000000   4918.000000   

           campaign         pdays      previous  emp.var.rate      latitude  \
count  43000.000000  43000.000000  43000.000000  43000.000000  43000.000000   
mean       2.567233    962.330953      0.174023      0.077128     36.856697   
std        2.772294    187.260394      0.497366      1.573898      7.225948   
min        1.000000      0.000000      0.000000     -3.400000     24.396000   
25%        1.000000    999.000000      0.000000     -1.800000     30.614750   
50%        2.000000    999.000000      0.000000      1.100000     36.761000   
75%        3.000000    999.000000      0.000000      1.400000     43.113250   
max       56.000000    999.000000      7.000000      1.400000     49.384000   

          longitude  
count  43000.000000  
mean     -95.939067  
std       16.752282  
min     -124.997000  
25%     -110.494250  
50%      -95.899500  
75%      -81.427750  
max      -66.937000

In [33]:
# podemos usar el método '.T' para transponer los resultados
df.describe().T

count        mean         std      min        25%       50%  \
age           37880.0   39.977112   10.437957   17.000   32.00000   38.0000   
default       34019.0    0.000088    0.009390    0.000    0.00000    0.0000   
housing       41974.0    0.535998    0.498708    0.000    0.00000    1.0000   
loan          41974.0    0.155620    0.362499    0.000    0.00000    0.0000   
duration      43000.0  257.739279  258.666033    0.000  102.00000  179.0000   
campaign      43000.0    2.567233    2.772294    1.000    1.00000    2.0000   
pdays         43000.0  962.330953  187.260394    0.000  999.00000  999.0000   
previous      43000.0    0.174023    0.497366    0.000    0.00000    0.0000   
emp.var.rate  43000.0    0.077128    1.573898   -3.400   -1.80000    1.1000   
latitude      43000.0   36.856697    7.225948   24.396   30.61475   36.7610   
longitude     43000.0  -95.939067   16.752282 -124.997 -110.49425  -95.8995   

                    75%       max  
age            47.00000    98.000  
default         0.00000     1.000  
housing         1.00000     1.000  
loan            0.00000     1.000  
duration      319.00000  4918.000  
campaign        3.00000    56.000  
pdays         999.00000   999.000  
previous        0.00000     7.000  
emp.var.rate    1.40000     1.400  
latitude       43.11325    49.384  
longitude     -81.42775   -66.937

**¿Qué es lo que vemos aquí?**

Entendamos los resultados de algunas de lascolumna `age`: 

- `count`: El número de observaciones no nulas en la columna "age".

- `mean`: La media de las edades en la columna.

- `std`: La desviación estándar, que mide la dispersión de las edades.

- `min`: La edad mínima en el conjunto de datos.

- `25%`: El percentil 25, que representa el valor por debajo del cual se encuentra el 25% de las edades.

- `50%`: La mediana o percentil 50, que es el valor que divide el conjunto de datos en dos mitades iguales.

- `75%`: El percentil 75, que representa el valor por debajo del cual se encuentra el 75% de las edades.

- `max`: La edad máxima en el conjunto de datos.



In [34]:
# hagamos lo mismo para las columnas categóricas, para eso tendremos que incluir entre los paréntesis el parámetro include = "object"
df.describe(include = "object").T

count unique                                   top   freq
job             42655     11                                admin.  10873
marital         42915      3                               MARRIED  25999
education       41193      7                     university.degree  12722
contact         43000      2                              cellular  27396
poutcome        43000      3                           NONEXISTENT  37103
cons.price.idx  42529     26                                93,994   7938
cons.conf.idx   43000     26                                 -36,4   8020
euribor3m       33744    309                                 4,857   2287
nr.employed     43000     11                                5228,1  16980
y               43000      2                                    no  38156
date            42752   1860                     4-septiembre-2019     41
id_             43000  43000  993bbbd6-4dbc-4a40-a408-f91f8462bee6      1

**Entendamos los resultados para la columna `job`**


- `count`: El número de observaciones no nulas en la columna "job".

- `unique`: La cantidad de valores únicos en la columna "job".

- `top`: El valor más común en la columna "job" (en este caso, "admin.").

- `freq`: La frecuencia del valor más común (en este caso, "admin.").


En el análisis preliminar de datos, existen dos métodos muy interesantes: `unique()` y `.value_counts()`, los cuales ya hemos aprendido en el módulo 2.

El método `unique()` se utiliza para obtener los valores únicos de una columna. Devuelve un *array* de los valores distintos presentes en esa columna, sin repetir. Será útil para identificar las categorías únicas en una variable categórica o los diferentes elementos en una columna de texto. 

Por otro lado, el método `.value_counts()` nos permite contar la frecuencia de cada valor en una columna. Devuelve una serie que muestra el valor como índice y el recuento como valor. Es útil para obtener información sobre la distribución de los valores en una columna, especialmente en variables categóricas. 

Ambos métodos son herramientas poderosas para explorar y comprender la naturaleza de los datos, y nos permiten obtener información clave sobre la distribución y la unicidad de los valores en nuestras variables.

In [13]:
# recordemos el DataFrame
df.head(1)

age        job  marital education  default  housing  loan    contact  \
0  NaN  housemaid  MARRIED  basic.4y      0.0      0.0   0.0  telephone   

   duration  campaign  pdays  previous     poutcome  emp.var.rate  \
0       261         1    999         0  NONEXISTENT           1.1   

  cons.price.idx cons.conf.idx euribor3m nr.employed   y           date  \
0         93,994         -36,4     4,857        5191  no  2-agosto-2019   

   latitude  longitude                                   id_  
0    41.495    -71.233  089b39d8-e4d0-461b-87d4-814d71e0e079

In [14]:
# saquemos los valores únicos para la columna de `marital` usando el método 'unique()'
df["marital"].unique()

array(['MARRIED', 'SINGLE', 'DIVORCED', nan], dtype=object)

In [15]:
# y si usamos el '.value_counts()' 
# en este caso vemos que dentro de nuestro conjunto de datos tenemos 26105 casados, 12147 solteros y 4833 divorciados
df["marital"].value_counts()

marital
MARRIED     25999
SINGLE      12105
DIVORCED     4811
Name: count, dtype: int64

Genial!! Pero solo lo hemos hecho para una columna, en realidad no interesaría hacerlo para todas las variables categóricas del DataFrame, ya que lo que en realidad queremos es explorar todas las columnas categóricas. Por lo tanto, lo que tendremos que hacer es: 

- Seleccionar las variables que son categóricas de nuestro DataFrame, para esto usaremos el método `.select_dtypes()` que es una función en Pandas que nos permite seleccionar columnas de un DataFrame basándose en su tipo de datos. Devuelve un DataFrame que contiene solo las columnas que cumplen con los tipos de datos especificados en los argumentos `include` y `exclude`. La sintaxis básica del método `select_dtypes()` es la siguiente:

    ```python
    DataFrame.select_dtypes(include=None, exclude=None)
    ```

    - `include` (opcional): Especifica los tipos de datos que se deben incluir en la selección. Puede ser una cadena de texto o una lista de tipos de datos, como `'int'`, `'float'`, `'object'`, etc. Si no se proporciona este argumento, se seleccionarán todas las columnas que coincidan con los tipos de datos especificados en `exclude`.

    - `exclude` (opcional): Especifica los tipos de datos que se deben excluir de la selección. Puede ser una cadena de texto o una lista de tipos de datos. Si no se proporciona este argumento, se seleccionarán todas las columnas que coincidan con los tipos de datos especificados en `include`.


    Algunos ejemplos de cómo se puede utilizar el método `select_dtypes()`:

    ```python
    # Seleccionar todas las columnas numéricas
    df.select_dtypes(include=['int', 'float'])

    # Seleccionar todas las columnas de tipo objeto (cadenas de texto)
    df.select_dtypes(include='object')

    # Seleccionar todas las columnas excepto las numéricas
    df.select_dtypes(exclude=['int', 'float'])

    # Seleccionar todas las columnas excepto las de tipo objeto (cadenas de texto)
    df.select_dtypes(exclude='object')
    ```

- Sacar los nombres de las columnas de tipo categórico. 

- Hacer un *for loop* para que nos haga los `.unique()` y los `.value_counts()` de cada columna

In [16]:
# sacamos las columnas que son de tipo object (categóricas) de nuestro DataFrame usando el método '.select_dtypes()'
df_cat = df.select_dtypes(include = "object")
df_cat.head()

job  marital    education    contact     poutcome cons.price.idx  \
0  housemaid  MARRIED     basic.4y  telephone  NONEXISTENT         93,994   
1   services  MARRIED  high.school  telephone  NONEXISTENT         93,994   
2   services  MARRIED  high.school  telephone  NONEXISTENT         93,994   
3     admin.  MARRIED     basic.6y  telephone  NONEXISTENT         93,994   
4   services  MARRIED  high.school  telephone  NONEXISTENT         93,994   

  cons.conf.idx euribor3m nr.employed   y                date  \
0         -36,4     4,857        5191  no       2-agosto-2019   
1         -36,4       NaN        5191  no  14-septiembre-2016   
2         -36,4     4,857        5191  no     15-febrero-2019   
3         -36,4       NaN        5191  no   29-noviembre-2015   
4         -36,4       NaN        5191  no       29-enero-2017   

                                    id_  
0  089b39d8-e4d0-461b-87d4-814d71e0e079  
1  e9d37224-cb6f-4942-98d7-46672963d097  
2  3f9f49b5-e410-4948-bf6e-f9244f04918b  
3  9991fafb-4447-451a-8be2-b0df6098d13e  
4  eca60b76-70b6-4077-80ba-bc52e8ebb0eb

Si miramos las columnas seleccionadas por Pandas nos damos cuenta que la mayoría tienen sentido, pero... que pasa con las columnas de 'ons.price.idx', 'cons.conf.idx', 'euribor3m', 'nr.employed' y 'date'. Si excluimos la columna de 'date', vemos claramente que son números, ¿qué esta pasando? Y es que, recordad que en Python los decimales se definen con ".", al ir estos números con comas automáticamente nos los convierte en *strings*. Esto ya aprenderemos a cambiarlo en próximas lecciones, pero ... ¿Cómo podríamos eliminar esas columnas para esta parte del análisis ya que no son columnas categóricas? Para eso usaremos el método `.drop()`. Este método se utiliza para eliminar filas o columnas de un DataFrame. Tiene la siguiente sintaxis básica:

```python
DataFrame.drop(labels, axis=0, inplace=False)
```

- `labels`: Especifica los nombres de las filas o columnas que se queremos eliminar. Puede ser un solo valor o una lista de valores.

- `axis`: Indica si se eliminarán filas (`axis=0`) o columnas (`axis=1`).

- `inplace` (opcional): Es un parámetro opcional que indica si se realizará la eliminación directamente en el DataFrame original (`inplace=True`) o si se creará un nuevo DataFrame con los cambios (`inplace=False`, valor por defecto).

En nuestro caso eliminaremos las columnas de 'ons.price.idx', 'cons.conf.idx', 'euribor3m', 'nr.employed' y 'date'. 


In [17]:
df_cat.head()

job  marital    education    contact     poutcome cons.price.idx  \
0  housemaid  MARRIED     basic.4y  telephone  NONEXISTENT         93,994   
1   services  MARRIED  high.school  telephone  NONEXISTENT         93,994   
2   services  MARRIED  high.school  telephone  NONEXISTENT         93,994   
3     admin.  MARRIED     basic.6y  telephone  NONEXISTENT         93,994   
4   services  MARRIED  high.school  telephone  NONEXISTENT         93,994   

  cons.conf.idx euribor3m nr.employed   y                date  \
0         -36,4     4,857        5191  no       2-agosto-2019   
1         -36,4       NaN        5191  no  14-septiembre-2016   
2         -36,4     4,857        5191  no     15-febrero-2019   
3         -36,4       NaN        5191  no   29-noviembre-2015   
4         -36,4       NaN        5191  no       29-enero-2017   

                                    id_  
0  089b39d8-e4d0-461b-87d4-814d71e0e079  
1  e9d37224-cb6f-4942-98d7-46672963d097  
2  3f9f49b5-e410-4948-bf6e-f9244f04918b  
3  9991fafb-4447-451a-8be2-b0df6098d13e  
4  eca60b76-70b6-4077-80ba-bc52e8ebb0eb

In [18]:
# eliminamos las columnas que no son de nuestro interés usando el método '.drop()'
# pasándolo una lista con las columnas que queremos eliminar
# axis = 1 porque los nombres que le hemos pasado en la lista corresponden a las columnas
# inplace = True porque queremos que los cambios realizados se sobreescriban en el DataFrame
df_cat.drop(['cons.price.idx', 'cons.conf.idx', 'euribor3m', 'nr.employed' , 'date'], axis = 1, inplace=True)

# chuequeamos como se ha quedado el DataFrame y vemos que ya lo tenemos preparado para poder evaluar cuales son sus valores únicos ('.unique()') y sus frecuencias ('.value_counts()')
df_cat.head()

job  marital    education    contact     poutcome   y  \
0  housemaid  MARRIED     basic.4y  telephone  NONEXISTENT  no   
1   services  MARRIED  high.school  telephone  NONEXISTENT  no   
2   services  MARRIED  high.school  telephone  NONEXISTENT  no   
3     admin.  MARRIED     basic.6y  telephone  NONEXISTENT  no   
4   services  MARRIED  high.school  telephone  NONEXISTENT  no   

                                    id_  
0  089b39d8-e4d0-461b-87d4-814d71e0e079  
1  e9d37224-cb6f-4942-98d7-46672963d097  
2  3f9f49b5-e410-4948-bf6e-f9244f04918b  
3  9991fafb-4447-451a-8be2-b0df6098d13e  
4  eca60b76-70b6-4077-80ba-bc52e8ebb0eb

In [19]:
# creamos una variable con los nombres de las columnas del DataFrame de las variables categóricas utilizando el método '.columns'
columnas_cat = df_cat.columns
print(f"Las columnas del DataFrame de variables categóricas son {columnas_cat}")

# empezamos a iterar por cada una de las columnas para sacar sus valores únicos y sus frecuencias
for columna in columnas_cat:
    print(f" \n----------- ESTAMOS ANALIZANDO LA COLUMNA: '{columna.upper()}' -----------\n")
    print(f"Sus valores únicos son: {df_cat[columna].unique()}\n")
    print(f"Las frecuencias de los valores únicos de las categorías son: {df_cat[columna].value_counts()} ")

Las columnas del DataFrame de variables categóricas son Index(['job', 'marital', 'education', 'contact', 'poutcome', 'y', 'id_'], dtype='object')
 
----------- ESTAMOS ANALIZANDO LA COLUMNA: 'JOB' -----------

Sus valores únicos son: ['housemaid' 'services' 'admin.' 'blue-collar' 'technician' 'retired'
 'management' 'unemployed' 'self-employed' nan 'entrepreneur' 'student']

Las frecuencias de los valores únicos de las categorías son: job
admin.           10873
blue-collar       9654
technician        7026
services          4162
management        3050
retired           1790
entrepreneur      1522
self-employed     1489
housemaid         1123
unemployed        1063
student            903
Name: count, dtype: int64 
 
----------- ESTAMOS ANALIZANDO LA COLUMNA: 'MARITAL' -----------

Sus valores únicos son: ['MARRIED' 'SINGLE' 'DIVORCED' nan]

Las frecuencias de los valores únicos de las categorías son: marital
MARRIED     25999
SINGLE      12105
DIVORCED     4811
Name: count, dtype: int64

Lo siguiente que deberíamos hacer es: 
- Tipos de variables que tenemos

- ¿Tenemos valores nulos?

- ¿Tenemos valores duplicados?

Manos a la obra!

## Valores nulos

Los valores nulos, también conocidos como valores faltantes o *missing values*, son aquellos que no están presentes en una variable o columna específica de un conjunto de datos. Estos valores pueden surgir por diversas razones, como errores en la recopilación de datos, problemas técnicos, falta de respuesta o simplemente porque algunos elementos no tienen valores registrados para ciertas variables.

Es importante hacer una buena identificación/gestión de los valores nulos por varias razones:

1. **Impacto en el análisis**: Los valores nulos pueden afectar negativamente el análisis de datos, ya que la mayoría de las técnicas y algoritmos requieren que todas las observaciones tengan valores completos. Si no se manejan adecuadamente, los valores nulos pueden generar resultados incorrectos o sesgados. Imaginad que estamos analizando los datos de ventas de una tienda minorista. Si no gestionas adecuadamente los valores nulos en la columna de "cantidad de productos vendidos", podríamos obtener resultados incorrectos al calcular la cantidad total de ventas, calcular el promedio de ventas diarias o realizar cualquier otro tipo de análisis. Si ignoramos los valores nulos, el análisis podría subestimar las ventas o sesgar los resultados, lo que llevaría a decisiones comerciales erróneas.

2. **Precisión y confiabilidad**: La presencia de valores nulos puede comprometer la precisión y confiabilidad de los resultados y conclusiones obtenidos a partir de los datos. Si se ignoran los valores nulos, se corre el riesgo de obtener estimaciones inexactas o interpretaciones incorrectas. Supongamos que estamos analizando datos de encuestas de satisfacción del cliente. Si no manejas los valores nulos en las respuestas de los clientes, podrías obtener estimaciones inexactas de la satisfacción general, ya que no tendrías en cuenta las opiniones de aquellos clientes que no respondieron ciertas preguntas. Esto podría llevar a conclusiones incorrectas sobre los aspectos que influyen en la satisfacción del cliente y, en última instancia, afectar la toma de decisiones para mejorar la calidad del servicio o producto.

3. **Integridad de los datos**: La gestión adecuada de los valores nulos ayuda a mantener la integridad general de los datos. Al identificar y manejar los valores nulos, se asegura que los datos sean coherentes y completos, lo que facilita su uso en análisis posteriores. Imaginemos que tenemos un conjunto de datos que contiene información sobre pacientes en un estudio clínico. Si no gestionaramos los valores nulos en la columna de "edad del paciente", podría haber inconsistencias en el análisis, ya que los valores nulos podrían afectar la validez de los cálculos basados en la edad, como la edad promedio del grupo de pacientes o las tasas de éxito del tratamiento según la edad. Al identificar y manejar adecuadamente los valores nulos, se garantiza que los datos sean coherentes y completos, lo que mejora la calidad y la fiabilidad de los análisis y las conclusiones.



En Python hay 5 tipos principales de valores nulos:

1. **NaN**: NaN, que significa "Not a Number", es el valor nulo más común en Pandas. Representa un valor faltante o no válido en una serie o columna de datos. En Pandas, NaN se considera como un valor de tipo float.

2. **None**: None es un objeto de Python que se utiliza para representar la ausencia de valor en un contexto no numérico. A diferencia de NaN, que se utiliza principalmente en contextos numéricos, None puede ser utilizado en cualquier tipo de dato en Pandas, incluyendo objetos, cadenas de texto, entre otros. None se considera un valor de tipo NoneType.

3. **valores texto**: a veces nos podemos encontrar con valores de tipo texto pero que no son valores reales, si no que nos indican que no hay valor. Algunos ejemplos de esto son: 
   
    -  `"n/a"`, `"NaN"`, `"nan"`, `"null"`: son NaN que serán literalmente *strings*. Esto puede ocurrir cuando se han metido en nuestra base de datos literalmente con esa estructura. Aún con todo, podemos vernos en dos situaciones: 
4. **Los 99999 o 00000**: es similar al caso anterior, pero en este caso vienen codificados como *integers*. Esto es usual encontrarlo en datos que vienen de sensores o sistemas de medición continua. 

5. **NaT**: los veremos cuando tengamos valores nulos en columnas de tipo `datetime`.

Una indentificación/gestión adecuada de los valores nulos implica tomar decisiones sobre cómo manejarlos. Este apartado lo no desarrollaremos ampliamente ya que tendremos una lección dedicada a la gestión de nulos en Pandas. Sin embargo, algunas de las estrategias más comunes que veremos son:

- Eliminación de filas o columnas con valores nulos: Si los valores nulos son pocos y no afectan significativamente el análisis, se pueden eliminar las filas o columnas correspondientes. Sin embargo, esta estrategia puede llevar a una pérdida de información si los valores nulos contienen información relevante.

- Imputación de valores: Consiste en estimar o reemplazar los valores nulos con valores calculados a partir de otros datos o utilizando técnicas de imputación, como el promedio, la mediana o modelos de imputación más complejos.

- Creación de variables indicadoras: En algunos casos, se pueden crear variables indicadoras o banderas que indiquen si un valor es nulo o no. 


### Métodos para identificar nulos en Pandas

En Pandas, puedes identificar los valores nulos en un DataFrame utilizando diferentes métodos y funciones. Aquí os presentamos las formas comunes de hacerlo:

1. **isnull()**: devuelve un DataFrame o una Serie booleana que indica si cada valor en el DataFrame es nulo o no. Los valores nulos se representan como `True`, mientras que los valores no nulos se representan como `False`.Se puede aplicar este método a un DataFrame completo o a una columna específica. Por ejemplo:

    ```python
    # Verificar si hay valores nulos en todo el DataFrame
    df.isnull()

    # Verificar si hay valores nulos en una columna específica
    df['columna'].isnull()
    ```

2. **notnull()**: Similar a `isnull()`, devuelve un DataFrame o una Serie booleana que indica si cada valor en el DataFrame no es nulo. Los valores no nulos se representan como `True` y los valores nulos como `False`. Al igual que `isnull()`, puedes aplicar este método a un DataFrame completo o a una columna específica.

    ```python
    # Verificar si hay valores no nulos en todo el DataFrame
    df.notnull()

    # Verificar si hay valores no nulos en una columna específica
    df['columna'].notnull()
    ```

3. **info()**: proporciona un resumen conciso de un DataFrame, incluyendo el número de valores no nulos en cada columna. Esto puede ser útil para identificar rápidamente las columnas que tienen valores nulos. Por ejemplo:

    ```python
    # Mostrar información resumida del DataFrame
    df.info()
    ```

4. **sum()**: Al llamar al método `sum()` en un DataFrame o una Serie booleana (como los que nos devuelve el método `.isnull()` o `.notnull()`), se puede obtener el recuento total de valores nulos en cada columna. El valor nulo se trata como `True` y se suma como 1, mientras que los valores no nulos se tratan como `False` y se suman como 0. Por ejemplo:

    ```python
    # Contar los valores nulos en cada columna del DataFrame
    df.isnull().sum()

    # Contar los valores nulos en una columna específica
    df['columna'].isnull().sum()
    ```
Pongámoslo en práctica!

In [20]:
# recordemos el DataFrame
df.head(1)

age        job  marital education  default  housing  loan    contact  \
0  NaN  housemaid  MARRIED  basic.4y      0.0      0.0   0.0  telephone   

   duration  campaign  pdays  previous     poutcome  emp.var.rate  \
0       261         1    999         0  NONEXISTENT           1.1   

  cons.price.idx cons.conf.idx euribor3m nr.employed   y           date  \
0         93,994         -36,4     4,857        5191  no  2-agosto-2019   

   latitude  longitude                                   id_  
0    41.495    -71.233  089b39d8-e4d0-461b-87d4-814d71e0e079

In [21]:
# ¿Cuántos valores nulos tenemos en la columna 'default' ?
# en este caso usaremos el método '.isnull()'
# como vemos nos devuelve una Serie que recordemos es una estructura de datos unidimensional que contiene datos etiquetados y se asemeja a una columna de un DataFrame
df["default"].isnull()

0        False
1         True
2        False
3        False
4        False
         ...  
19154    False
26206    False
15046    False
15280     True
27570    False
Name: default, Length: 43000, dtype: bool

In [22]:
# lo que nos devuelve el método '.isnull()' por si mismo es poco entendible, 
# es por esto que nos suma todos los valores nulos que ha indentificado como False usando el método '.sum()'
# en este caso tenemos 9016 valores nulos para la columna 'default'. 
df["default"].isnull().sum()

8981

In [23]:
# en ejemplo anterior hemos visto el número de nulos para una sola columna,
# en este ejemplo lo haremos para todo el DataFrame, para que nos diga el número de nulos que hay por columna
df.isnull().sum()

age               5120
job                345
marital             85
education         1807
default           8981
housing           1026
loan              1026
contact              0
duration             0
campaign             0
pdays                0
previous             0
poutcome             0
emp.var.rate         0
cons.price.idx     471
cons.conf.idx        0
euribor3m         9256
nr.employed          0
y                    0
date               248
latitude             0
longitude            0
id_                  0
dtype: int64

En este ejemplo vemos que solo tenemos valores nulos para algunas columnas, en concreto para:

- age, la cual tiene 5120 nulos. 

- job, la cual tiene 345 nulos. 

- marital, la cual tiene 85 nulos. 

- education, la cual tiene 1807 nulos. 

- default, la cual tiene 8981 nulos.

- housing, la cual tiene 1026 nulos.

- loan, la cual tiene 1026 nulos, 

- cons.price.idx, la cual tienne 471 nulos. 

- euribor3m, la cual tiene 9256 nulos.

- date, la cual tiene 248 nulos. 


🚨 **NOTA** En Pandas existe el método `.isna()` que es exactamente igual que el `.isnull()`. Su sintaxis básica es:

```python
dataframe.isna()
# fijaos como es exactamente igual que el '.isnull()`
```

In [24]:
# veamos ahora el mismo ejemplo pero usando el méetodo '.notnull()`
# que recordemos nos devuelve el número de valores no nulos que tenemos en cada columna
# en este caso, nos esta diciendo que en la columna "default" tenemos 34154 valores no nulos, es decir, 34154 con valores. 
df["default"].notnull().sum()

34019

In [25]:
# si lo hacemos para todo el DataFrame
df.notnull().sum()

age               37880
job               42655
marital           42915
education         41193
default           34019
housing           41974
loan              41974
contact           43000
duration          43000
campaign          43000
pdays             43000
previous          43000
poutcome          43000
emp.var.rate      43000
cons.price.idx    42529
cons.conf.idx     43000
euribor3m         33744
nr.employed       43000
y                 43000
date              42752
latitude          43000
longitude         43000
id_               43000
dtype: int64

In [26]:
# hemos comentado que también podemos usar el método .info()' 
# que recordemos muestra información resumida sobre el DataFrame, incluyendo el número de datos, tipos de datos o valores no nulos
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 43000 entries, 0 to 27570
Data columns (total 23 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   age             37880 non-null  float64
 1   job             42655 non-null  object 
 2   marital         42915 non-null  object 
 3   education       41193 non-null  object 
 4   default         34019 non-null  float64
 5   housing         41974 non-null  float64
 6   loan            41974 non-null  float64
 7   contact         43000 non-null  object 
 8   duration        43000 non-null  int64  
 9   campaign        43000 non-null  int64  
 10  pdays           43000 non-null  int64  
 11  previous        43000 non-null  int64  
 12  poutcome        43000 non-null  object 
 13  emp.var.rate    43000 non-null  float64
 14  cons.price.idx  42529 non-null  object 
 15  cons.conf.idx   43000 non-null  object 
 16  euribor3m       33744 non-null  object 
 17  nr.employed     43000 non-null  obje

Si interpretaramos los resultados obtenidos del método `.info()` que:

- Nos dice el número de filas que tenemos en el DataFrame, en este caso 43000

- Nos dice el número total de columnas que tenemos en el DataFrame, en este caso 23

- Los nombres de todas las columnas del DataFrame, en este caso 'age', 'job', 'marital', 'education', 'default', 'housing', 'loan', 'contact', 'duration', 'campaign', 'pdays', 'previous', 'poutcome', 'emp.var.rate', 'cons.price.idx', 'cons.conf.idx', 'euribor3m', 'nr.employed', 'y', 'month_day_week'

- El tipo de datos que tenemos en nuestro DataFrame, con el número de columnas que tenemos de cada tipo, en este caso float64(4), int64(5), object(11)

- Para cada una de las columnas tenemos:

    - El número de valores no nulos. En el caso de la columna `job` tiene 42655  valores no nulos

    - El tipo de la columna. En el caso de la columna `job` es de tipo object, que recordemos que es como el *string* con el que estamos familiarizadas.   

## Valores duplicados

Los valores duplicados se refieren a la presencia de registros/filas repetidos en un conjunto de datos. Esto significa que hay al menos una o más observaciones o filas que comparten exactamente los mismos valores en todas las columnas.

Existen varias razones por las cuales podemos tener valores duplicados en un conjunto de datos. Algunas de ellas son:

1. **Errores de entrada**: Durante la recopilación o inserción de datos, pueden ocurrir errores humanos o técnicos, lo que resulta en la duplicación involuntaria de registros.

2. **Fusión de conjuntos de datos**: Cuando se combinan diferentes fuentes de datos, puede haber duplicados entre los conjuntos originales, especialmente si no se lleva a cabo un proceso adecuado de eliminación de duplicados durante la fusión.

3. **Actualizaciones o modificaciones**: Si se realizan actualizaciones o modificaciones en un conjunto de datos existente, es posible que los valores duplicados se creen debido a cambios en los registros.

La gestión adecuada de los valores duplicados es esencial en el proceso de análisis de datos por varias razones:

1. **Precisión y validez de los resultados**: Los valores duplicados pueden afectar la precisión de los análisis y generar resultados sesgados o inexactos. Al eliminar los duplicados, nos aseguramos de trabajar con datos únicos y obtener resultados más confiables.  Imaginemos que estamos realizando un estudio sobre preferencias de consumo de alimentos en una encuesta. Si permitimos que los participantes envíen respuestas duplicadas, podrías obtener una imagen distorsionada de las preferencias, ya que algunos alimentos podrían parecer más populares de lo que realmente son. Al eliminar los duplicados, puedes obtener resultados más precisos y confiables para comprender las preferencias reales.

2. **Eficiencia en el procesamiento**: Los valores duplicados aumentan el tamaño y la complejidad del conjunto de datos, lo que puede ralentizar el procesamiento y consumo de recursos innecesarios. Al eliminar los duplicados, optimizamos el rendimiento y la eficiencia del análisis.  Supongamos que estamos analizando registros de ventas en una empresa de comercio electrónico con una gran cantidad de datos. Si hay valores duplicados en los registros de ventas, el procesamiento de esos datos llevará más tiempo y requerirá más recursos computacionales. Al eliminar los duplicados, optimizas el proceso de análisis, lo que conduce a un procesamiento más rápido y eficiente. 

3. **Consistencia y coherencia de los datos**: La presencia de valores duplicados puede generar inconsistencias en los datos y dificultar la interpretación correcta. Al gestionar los duplicados, garantizamos la coherencia de los datos y facilitamos su interpretación. Consideremos ahora una base de datos de empleados que incluye información personal, como nombres y direcciones. Si hay registros duplicados en la base de datos debido a errores de entrada, se puede generar inconsistencia en los datos. Por ejemplo, un empleado puede tener múltiples entradas con diferentes direcciones. Al gestionar los duplicados, se asegura la coherencia de los datos y se evita la confusión al interpretar la información correcta de cada empleado.

4. **Toma de decisiones correctas**: Los análisis se realizan con el propósito de tomar decisiones informadas. Los valores duplicados pueden distorsionar la información y llevar a decisiones erróneas. Al gestionar los duplicados, mejoramos la calidad de los datos y, por ende, la toma de decisiones. Supongamos que estamos realizando un análisis de satisfacción del cliente y deseamos identificar los problemas más frecuentes reportados. Si hay valores duplicados en las quejas registradas, es posible que los problemas se vean amplificados y se tomen decisiones basadas en información distorsionada. Al gestionar los duplicados, puedes obtener una visión más precisa y equilibrada de los problemas reales, lo que facilita la toma de decisiones correctas para abordarlos de manera efectiva.

¿Qué haremos con los duplicados cuando los tengamos? Aquí os mostramos algunas de las opciones, pero al igual que en el caso de los nulos, lo desarrollaremos apliamente más adelante en el bootcamp:

-  **Eliminar los duplicados**: Si los duplicados no son relevantes para el análisis y no aportan información adicional. 

-  **Agrupar y resumir**: En algunos casos, puede ser útil agrupar los datos por ciertas columnas y realizar resúmenes o agregaciones sobre los valores duplicados.

-  **Marcar y analizar**: En lugar de eliminar los duplicados, podemos agregar una columna adicional para marcar los registros duplicados y luego analizarlos por separado. 

- **Investigar la causa**: Si los duplicados son inesperados o indican posibles problemas en los datos de origen, es importante investigar y corregir la causa del problema. Esto puede implicar verificar los procesos de entrada de datos, la calidad de los datos de origen o solucionar cualquier problema técnico que pueda estar generando duplicados.


### Métodos para indentificar los duplicados en Pandas

En Pandas, solo podemos identificar los duplicados usando el método `.duplicated()` el cual va a identificar las filas duplicadas en un DataFrame y devolverá una Serie booleana que indica True para las filas duplicadas y False para las filas únicas.

Su sintaxis básica es:

```python
df.duplicated(subset=['columna1', 'columna2'], keep='first')
```

El método `duplicated()` tiene los siguientes parámetros principales:

- **subset** (opcional): Este parámetro permite especificar las columnas sobre las cuales se comprobarán los duplicados. Puedes pasar una lista de nombres de columnas para considerar solo esas columnas al buscar duplicados. Si no se especifica, se considerarán todas las columnas del DataFrame.

- **keep** (opcional): Este parámetro controla qué ocurrencias de los duplicados se marcarán como `True`. Los valores posibles son:

  - `'first'` (valor predeterminado): Marca como `True` todas las filas duplicadas, excepto la primera aparición de cada conjunto de duplicados.

  - `'last'`: Marca como `True` todas las filas duplicadas, excepto la última aparición de cada conjunto de duplicados.
  
  - `False`: Marca como `True` todas las filas duplicadas, incluyendo todas las apariciones de duplicados.


In [27]:
# recordemos el DataFrame, mostrando dos filas aleatorias
df.sample(2)

age          job   marital education  default  housing  loan  \
38785  56.0  blue-collar  DIVORCED  basic.4y      0.0      0.0   0.0   
29384  45.0     services    SINGLE       NaN      0.0      0.0   0.0   

         contact  duration  campaign  pdays  previous     poutcome  \
38785   cellular       134         1    999         0  NONEXISTENT   
29384  telephone       541         1    999         0  NONEXISTENT   

       emp.var.rate cons.price.idx cons.conf.idx euribor3m nr.employed    y  \
38785          -3.4         92,649         -30,1       NaN      5017,5  yes   
29384          -1.8         93,075         -47,1     1,405      5099,1  yes   

                  date  latitude  longitude  \
38785  27-febrero-2019    33.486    -85.642   
29384     8-abril-2019    39.286   -112.989   

                                        id_  
38785  ec796179-6e5d-4f02-845e-688f9c8e2689  
29384  6beaeefb-a937-4b00-a61d-5cbbbae5a999

In [28]:
# ¿tenemos duplicados teniendo en cuenta todas las columnas del DataFrame?
# al igual que en el caso del '.isnull()' nos devuelve una serie de Booleanos
df.duplicated()

0        False
1        False
2        False
3        False
4        False
         ...  
19154    False
26206    False
15046    False
15280    False
27570    False
Length: 43000, dtype: bool

In [29]:
# como hemos visto lo anterior nos devuelve una serie de booleanos lo cual es poco entendible, 
# por lo que añadiremos el método '.sum()' para que nos cuente cuantas filas tenemos duplicadas
# en este caso el código nos esta diciendo que tenemos 1994 filas duplicadas
df.duplicated().sum()

0

In [30]:
# ¿Y si quisieramos saber si tenemos duplicados para una columna en concreto? 
# en ese caso tendremos que usar el parámetro subset()' 
# imaginemos que queremos saber si hay duplicados en la columna de 'date' 
# en este caso tenemos un total de 41309 fechas duplicadas en nuestro conjunto de datos (recordemos que teníamos 43170 datos)

df.duplicated(subset = "date").sum()

41139

In [31]:
# si por ejemplo evaluaramos si existen valores duplicados para la columna de 'id_' debería salir que no existe ningún valor duplicado, ya que estos id deberían ser únicos
# usemos de nuevo el parámetro 'subset'
df.duplicated(subset  = "id_").sum() 

0

# Resumen de los conceptos aprendidos


| Método                       | Descripción                                           | Ejemplo                                              |
|------------------------------|-------------------------------------------------------|------------------------------------------------------|
| `pd.read_csv()`              | Cargar datos desde un archivo CSV en un DataFrame.  | `import pandas as pd`<br>`data = pd.read_csv('archivo.csv')` |
| `.head()`                    | Muestra las primeras filas de un DataFrame.        | `data.head()`                                        |
| `.tail()`                    | Muestra las últimas filas de un DataFrame.         | `data.tail()`                                        |
| `.sample()`                  | Devuelve filas aleatorias de un DataFrame.          | `data.sample(3)`                                     |
| `.shape`                     | Devuelve el número de filas y columnas.             | `data.shape`                                         |
| `.columns`                   | Muestra los nombres de las columnas.                | `data.columns`                                       |
| `unique()`                   | Devuelve valores únicos en una columna.             | `data['columna'].unique()`                          |
| `value_counts()`             | Cuenta la frecuencia de valores únicos en una columna. | `data['columna'].value_counts()`                   |
| `select_dtypes()`            | Selecciona columnas por tipo de datos.              | `data.select_dtypes(include=['int'])`               |
| `drop()`                     | Elimina filas o columnas de un DataFrame.          | `data.drop(columns=['columna_a_eliminar'], inplace=True)` |
| `.info()`                    | Muestra información sobre el DataFrame.            | `data.info()`                                        |
| `.isnull().sum()`            | Calcula la cantidad de valores nulos en cada columna. | `data.isnull().sum()`                             |
| `.notnull().sum()`           | Calcula la cantidad de valores no nulos en cada columna. | `data.notnull().sum()`                           |
| `.duplicated().sum()`        | Calcula la cantidad de filas duplicadas en el DataFrame. | `data.duplicated().sum()`                        |
| `describe()`                 | Proporciona estadísticas descriptivas del DataFrame. | `data.describe()` o `df.describe(include = "object")`                                  |




# Ejercicios

1- Abrid el fichero de `rating-and-performance` con el método que sea más correcto.

2- Mostrad las 10 últimas filas.

3- Mostrad 5 filas aleatorias del DataFrame.

4- ¿Cuántas filas y columnas tenemos en el DataFrame? 

5- ¿Cuáles son los nombres de las columnas del DataFrame?

6- Seleccionad solo las columnas de tipo categórico del DataFrame. Después, mostrad los valores únicos y sus frecuencias para cada una de las columnas seleccionadas. 

7- En el DataFrame tenemos dos columnas que parece que nos dan la misma información "title" y "title_orig". Eliminad "title". 

8- ¿En que columnas tenemos valores nulos? Utilizad el método que más os guste. 

9- ¿Tenemos valores duplicados en el DataFrame?